File: protobuf-serialization/src/test/java/com/stannis/benchmark/repository/RecipeRepositoryImplTest.java
package com.stannis.benchmark.repository;

import com.stannis.benchmark.mapper.RecipeProtoMapper;
import com.stannis.benchmark.model.Ingredient;
import com.stannis.benchmark.model.IngredientUsed;
import com.stannis.benchmark.model.MeasurementType;
import com.stannis.benchmark.model.Recipe;
import com.stannis.benchmark.util.RecipeTestUtil;
import com.google.common.collect.ImmutableList;
import generated.stannis.benchmark.Messages;

import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class RecipeRepositoryImplTest {

    private Recipe recipe;
    private Messages.Recipe protoRecipe;
    private String ingredientName = "Cheese";
    private String recipeName = "My Recipe";
    private int quantity = 4;

    private RecipeRepository repository;

    @Before
    public void setUp() throws Exception {
        IngredientUsed cheeseUsed = new IngredientUsed(new Ingredient(ingredientName, "Creamy Cheese"), MeasurementType.OUNCE, quantity);

        recipe = RecipeTestUtil.createRecipe(recipeName, "Some spicy recipe using a few items", ImmutableList.of(cheeseUsed));
        protoRecipe = RecipeProtoMapper.parseAsProto(recipe);

        repository = new RecipeRepositoryImpl();
    }

    @Test
    public void storeProtoBufRecipe() {
        Integer id = repository.save(protoRecipe);
        Integer second = repository.save(protoRecipe);

        assertThat(id, is(0));
        assertThat(second, is(1));
    }

    @Test
    public void getProtoBufRecipeById() {
        Integer id = repository.save(protoRecipe);

        Messages.Recipe foundById = repository.getProtoRecipeById(id);

        assertThat(foundById.getName(), is(protoRecipe.getName()));
        assertThat(foundById.getIngredientsCount(), is(1));

        Messages.IngredientUsed firstIngredient = foundById.getIngredients(0);
        assertThat(firstIngredient.getQuantity(), is(quantity));
        assertThat(firstIngredient.getIngredient().getName(), is(ingredientName));
    }


    @Test
    public void storeBothRecipes() {
        Integer rFirst = repository.save(recipe);
        Integer rSecond = repository.save(recipe);

        Integer bFirst = repository.save(protoRecipe);
        Integer bSecond = repository.save(protoRecipe);

        assertThat(rFirst, is(0));
        assertThat(rSecond, is(1));

        assertThat(bFirst, is(0));
        assertThat(bSecond, is(1));
    }

    @Test
    public void storeJavaRecipe() {
        Integer id = repository.save(recipe);
        Integer second = repository.save(recipe);

        assertThat(id, is(0));
        assertThat(second, is(1));
    }
}

File: protobuf-bench/src/test/kotlin/com/stannis/protobufbench/repository/RecipeRepositoryImplTest.kt
package com.stannis.protobufbench.repository

import com.google.common.collect.ImmutableList
import org.hamcrest.CoreMatchers


class RecipeRepositoryImplTest {
    private var recipe: Recipe? = null
    private var protoRecipe: Messages.Recipe? = null
    private val ingredientName = "Cheese"
    private val recipeName = "My Recipe"
    private val quantity = 4
    private var repository: RecipeRepository? = null
    @Before
    @Throws(Exception::class)
    fun setUp() {
        val cheeseUsed = IngredientUsed(Ingredient(ingredientName, "Creamy Cheese"), MeasurementType.OUNCE, quantity)
        recipe =
            RecipeTestUtil.createRecipe(recipeName, "Some spicy recipe using a few items", ImmutableList.of(cheeseUsed))
        protoRecipe = RecipeProtoMapper.parseAsProto(recipe)
        repository = RecipeRepositoryImpl()
    }

    @Test
    fun storeProtoBufRecipe() {
        val id: Int = repository.save(protoRecipe)
        val second: Int = repository.save(protoRecipe)
        assertThat(id, CoreMatchers.`is`(0))
        assertThat(second, CoreMatchers.`is`(1))
    }

    @get:Test
    val protoBufRecipeById: Unit
        get() {
            val id: Int = repository.save(protoRecipe)
            val foundById: Messages.Recipe? = repository!!.getProtoRecipeById(id)
            assertThat(foundById.getName(), `is`(protoRecipe.getName()))
            assertThat(foundById.getIngredientsCount(), CoreMatchers.`is`(1))
            val firstIngredient: Messages.IngredientUsed = foundById.getIngredients(0)
            assertThat(firstIngredient.getQuantity(), CoreMatchers.`is`(quantity))
            assertThat(firstIngredient.getIngredient().getName(), CoreMatchers.`is`(ingredientName))
        }

    @Test
    fun storeBothRecipes() {
        val rFirst: Int = repository.save(recipe)
        val rSecond: Int = repository.save(recipe)
        val bFirst: Int = repository.save(protoRecipe)
        val bSecond: Int = repository.save(protoRecipe)
        assertThat(rFirst, CoreMatchers.`is`(0))
        assertThat(rSecond, CoreMatchers.`is`(1))
        assertThat(bFirst, CoreMatchers.`is`(0))
        assertThat(bSecond, CoreMatchers.`is`(1))
    }

    @Test
    fun storeJavaRecipe() {
        val id: Int = repository.save(recipe)
        val second: Int = repository.save(recipe)
        assertThat(id, CoreMatchers.`is`(0))
        assertThat(second, CoreMatchers.`is`(1))
    }
}
